// @airiot/settings/src/settings/message/dingtalk.js
import React from 'react'
import { Button, message } from 'antd'
import { api } from 'xadmin'
import { Icon } from 'xadmin-ui'
import { ReloadOutlined } from '@ant-design/icons'
import locales from '../../../locales'
const loadDingtalkService = () => {
  api({ name: 'dingtalk/handler/reload' })
    .fetch('', {})
    .then(({ json }) => {
      message.success('重新加载成功')
    })
    .catch(err => {
      message.error(err.message)
    })
}

const loadCmdService = () => {
  api({ name: 'cmd/handler/reload' })
    .fetch('', {})
    .then(({ json }) => {
      message.success('重新加载成功')
    })
    .catch(err => {
      message.error(err.message)
    })
}

export default {
  name: 'iot-dingtalk',
  'settings.schema': app => (props, cb) => {
    cb(null, {
      name: 'dingtalk',
      title: _r('钉钉配置'),
      parentTab: _r('通知服务配置'),
      type: 'object',
      properties: {
        agentid: {
          type: 'string',
          title: _r('应用ID')
        },
        appKey: {
          type: 'string',
          title: _r('应用key')
        },
        appSecret: {
          type: 'string',
          title: _r('应用密钥'),
          fieldType: 'password',
        }
      },
      form: ['agentid', 'appKey', 'appSecret'],
      required: ['agentid', 'appKey', 'appSecret']
    })
  },
  locales,
}
