// @airiot/settings/src/settings/message/wechat.js

import locales from '../../../locales'

export default {
  name: 'iot-wechat',
  locales,
  'settings.schema': app => (props, cb) => {
    cb(null, {
      name: 'wechat',
      title: _r('微信公众号'),
      type: 'object',
      parentTab: '通知服务配置',
      properties: {
        appid: {
          type: 'string',
          title: _r('公众号ID')
        },
        secret: {
          type: 'string',
          title: _r('密钥'),
          fieldType: 'password',
        },
        template: {
          title: _r('微信模板'),
          type: 'array',
          items: {
            type: 'object',
            title: _r('微信模板'),
            properties: {
              title: {
                type: 'string',
                title: _r('标识名称'),
                field: {
                  validate: (value, values) => {
                    const templates = values?.template                  
                    return value && _.compact(templates?.map(t => t?.title == value))?.length > 1 ? '标识名称不允许重复' : null
                  }
                }
              },
              id: {
                type: 'string',
                title: _r('模板标识'),
                field: {
                  validate: (value, values) => {
                    const templates = values?.template                  
                    return value && _.compact(templates?.map(t => t?.id == value))?.length > 1 ? '模板标识不允许重复' : null
                  }
                }
              },
              content: {
                type: 'string',
                title: _r('模板内容'),
              }
            },
            required: ['id', 'title', 'content']
          }
        }
      },
      form: ['appid', 'secret', 'template'],
      required: ['appid', 'secret']
    })
  }
}
