// @airiot/echarts/src/charts/mix/index.js
import React from 'react'
import { C } from 'xadmin-ui'
import _ from 'lodash'
import useChart from '../../components/useChart'
import { initLayout, chartDafaultData, defaultSchema } from '../common/base'

import {
  seriesOption as seriesOption1,
  formToOption as formToOption1,
  optionToForm as optionToForm1
} from '../line'
import {
  seriesOption as seriesOption2,
  formToOption as formToOption2,
  optionToForm as optionToForm2
} from '../bar'
import {
  seriesOption as seriesOption3,
  formToOption as formToOption3,
  optionToForm as optionToForm3
} from '../pie'
import {
  seriesOption as seriesOption4
} from '../scatter'

const echartOption = {
  line: {
    seriesOption: seriesOption1,
    formToOption: formToOption1,
    optionToForm: optionToForm1
  },
  bar: {
    seriesOption: seriesOption2,
    formToOption: formToOption2,
    optionToForm: optionToForm2
  },
  pie: {
    seriesOption: seriesOption3,
    formToOption: formToOption3,
    optionToForm: optionToForm3
  },
  scatter: {
    seriesOption: seriesOption4
  }
}

const chart_mix = {
  title: _r('混合图'),
  category: [_r('数据图表'), _r('混合图')],
  icon: require('../../svg/newIcons/混合图.svg'),
  component: props => useChart({ ...props }),
  initLayout,
  initParam: {
    option: {
      series: [{ type: 'line' }, { type: 'bar' }],
      xAxis: { show: true, type: 'category', axisLine: { lineStyle: { color: '#333' } } },
      yAxis: { show: true, type: 'value', axisLine: { lineStyle: { color: '#333' } } },
      dataset: chartDafaultData
    }
  },
  echartOption,
  paramSchema: {
    type: 'object',
    properties: {
      ...defaultSchema
    }
  }
}

export default chart_mix
export { echartOption }
