// @airiot/components/src/front.js
import app from 'xadmin'
import DemoComponents from './Demo/components'
import DemoFieldTypes from './Demo/fieldTypes'
import DemoHooks from './Demo/hooks'
import SelectTableRecord from "./components/SelectTableRecord"
import SelectTable from "./components/SelectTable"
import SelectTableSchema from './components/SelectTableSchema'
import ModelPointTable from "./components/ModelPointTable"
import PointTable from "./components/PointTable"
import FiltersHistory from './components/FiltersHistory'
import QueryEditor from './components/QueryEditor'
import MenuEditTree from './components/MenuEditTree'
import TagWrap from './components/TagWrap'
import TabelSchemaTabs from './components/TabelSchemaTabs'
import { TableTreeWidgets } from './components/TableTree'
import hooks from './hooks'

import dashboard from './dashboard'
import fieldTypes from './fieldTypes'
import model from './model'
import TableTree from './components/TableTree'
import locales from '../locales'
import ClearData from './components/ClearData'

export default app
  .use({
    name: 'airiot.components',
    locales,
    components: {
      'Common.SelectTableRecord': SelectTableRecord,
      'Common.SelectTable': SelectTable,
      'Common.SelectTableSchema': SelectTableSchema,
      PointTable, MenuEditTree,
      ModelPointTable,
      FiltersHistory,
      QueryEditor,
      TagWrap,
      TabelSchemaTabs,
      'Common.ClearData': ClearData,
      ...model.components
    },
    routers: {
      '/app/' : [
        {
          path: 'components',
          breadcrumbName: '公共组件',
          component: DemoComponents
        },
        {
          path: 'fieldTypes',
          breadcrumbName: 'fieldTypes',
          component: DemoFieldTypes
        }
        // {
        //   path: 'hooks',
        //   breadcrumbName: 'hooks',
        //   component: DemoHooks
        // }
      ]
    },
    hooks,
    dashboardWidgets: {
      TableTreeWidgets
    }
  })
  .use(dashboard)
  .use(fieldTypes)
