// @airiot/echarts/src/charts/common/option.js
const markLine = {
  title: _r('标线'),
  type: 'object',
  properties: {
    data: {
      title: _r('标线'),
      type: 'array',
      items: {
        type: 'object',
        properties: {
          name: {
            title: _r('名称'),
            type: 'string'
          },
          type: {
            title: _r('类型'),
            type: 'string',
            enum: ['min', 'max', 'average', 'median'],
            enum_title: [_r('最小值'), _r('最大值'), _r('点虚线'), _r('平均值')]
          },
          xAxis: {
            title: _r('X轴值'),
            type: 'number'
          },
          yAxis: {
            title: _r('Y轴值'),
            type: 'number'
          },
          color: {
            title: _r('颜色'),
            type: 'string',
            fieldType: 'color'
          },
          width: {
            title: _r('线宽'),
            type: 'number'
          }
        }
      }
    }
  }
}

const shadowStyle = {
  shadowBlur: {
    type: 'number',
    title: _r('阴影模糊')
  },
  shadowColor: {
    type: 'string',
    title: _r('阴影颜色'),
    fieldType: 'color'
  },
  shadowOffsetX: {
    type: 'number',
    title: _r('阴影水平偏移')
  },
  shadowOffsetY: {
    type: 'number',
    title: _r('阴影垂直偏移')
  },
  opacity: {
    type: 'number',
    title: _r('阴影透明度'),
    fieldType: 'slider',
    maximum: 1
  }
}

const borderStyle = {
  borderColor: {
    type: 'string',
    title: _r('边框颜色'),
    fieldType: 'color',
  },
  borderWidth: {
    type: 'number',
    title: _r('边框线宽')
  },
  borderType: {
    type: 'string',
    title: _r('类型'),
    enum: ['solid', 'dashed', 'dotted'],
    enum_title: [_r('实线'), _r('虚线'), _r('点虚线')]
  },
  borderRadius: {
    type: 'number',
    title: _r('圆角半径')
  }
}

const lineStyle = {
  title: _r('线样式'),
  type: 'object',
  properties: {
    color: {
      title: _r('颜色'),
      type: 'string',
      fieldType: 'color'
    },
    width: {
      title: _r('宽度'),
      type: 'number'
    },
    type: {
      type: 'string',
      title: _r('类型'),
      enum: ['solid', 'dashed', 'dotted'],
      enum_title: [_r('实线'), _r('虚线'), _r('点虚线')]
    }
  }
}

const textCommonStyle = {
  type: 'object',
  properties: {
    color: {
      type: 'string',
      title: _r('文字颜色'),
      fieldType: 'color'
    },
    backgroundColor: {
      type: 'string',
      title: _r('背景颜色'),
      fieldType: 'color',
    },
    fontSize: {
      type: 'number',
      title: _r('文字大小')
    },
    fontWeight: {
      type: 'number',
      title: _r('字体粗细')
    },
    lineHeight: {
      type: 'number',
      title: _r('行高')
    },
    fontStyle: {
      type: 'string',
      title: _r('字体风格'),
      enum: ['normal', 'italic'],
      enum_title: [_r('正常'), _r('斜体')]
    },
    align: {
      type: 'string',
      title: _r('水平对齐'),
      enum: ['left', 'center', 'right'],
      enum_title: [_r('左对齐'), _r('居中'), _r('右对齐')]
    },
    verticalAlign: {
      type: 'string',
      title: _r('垂直对齐'),
      enum: ['top', 'middle', 'bottom'],
      enum_title: [_r('顶部对其'), _r('中间对齐'), _r('底部对齐')]
    }
  }
}

export { markLine, shadowStyle, borderStyle, lineStyle, textCommonStyle }
