// @airiot/data-api/src/AIalgorithm/index.js
import { C } from "xadmin-ui"
import ListPage from "./pages/ListPage"
import AddEditPage from "./pages/AddEditPage"
import MyListPage from "./pages/MyListPage"
import AlgorDebugPage from "./pages/AlgorDebugPage"
import ApiComponent from './components/ApiComponent'
import { InputParams, OutputParams } from "./components/ParamsShow"
import { ServiceSelect, serviceFunc } from "./components/service"
import './index.css'

export default {
  name: 'iot.algorithm',
  models: {
    // Algorithm: {
    //   name: 'algorithm',
    //   title: _r('AI算法库'),
    //   type: 'object',
    //   resource: 'ds/algorithm',
    //   properties: {
    //     name: {
    //       title: '算法名称',
    //       type: 'string'
    //     },
    //     key: {
    //       title: '算法标识',
    //       type: 'string'
    //     },
    //     describe: {
    //       title: '算法描述',
    //       type: 'string'
    //     },
    //     runMode: {
    //       title: '算法部署',
    //       type: 'string',
    //       enum: ['api', 'service'],
    //       enum_title: ['API接口', '服务']
    //     }
    //   },
    //   listFields: ['name', 'key', 'describe', 'runMode'],
    //   required: ['name', 'key', 'describe', 'runMode'],
    //   form: ['*'],
    //   permission: { add: true, edit: true, view: true, delete: true },
    //   components: {
    //     ListPage: ListPage
    //   }
    // },
    algorithm: {
      name: 'algorithm',
      title: _r('我的算法'),
      type: 'object',
      resource: 'algorithm/algorithm',
      icon: 'user',
      is_extend: true,
      properties: {
        name: {
          title: '算法名称',
          type: 'string'
        },
        key: {
          title: '算法标识',
          type: 'string'
        },
        describe: {
          title: '算法描述',
          type: 'string'
        },
        runMode: {
          title: '算法部署',
          type: 'string',
          enum: ['api', 'service'],
          enum_title: ['API接口', '服务'],
          field: {
            effect: ({ value }, form) => {
              setTimeout(() => {
                if (value === 'api') {
                  form.setFieldData('apiKey', { display: true })
                  form.setFieldData('algorithmID', { display: false })
                  form.setFieldData('function', { display: false })
                  form.change('algorithmID', null)
                  form.change('function', null)
                } else {
                  form.setFieldData('apiKey', { display: false })
                  form.setFieldData('algorithmID', { display: true })
                  form.setFieldData('function', { display: true })
                  form.change('apiKey', null)
                }
              })
            }
          }
        },
        apiKey: {
          title: '接口地址',
          type: 'string',
          field: {
            component: ApiComponent
          }
        },
        algorithmID: {
          title: '服务选择',
          type: 'string',
          field: {
            component: ServiceSelect
          }
        },
        function: {
          title: '函数选择',
          type: 'string',
          field: {
            component: serviceFunc
          }
        },
        inputParams: {
          title: _r('输入参数'),
          type: 'object',
          properties: {},
          field: {
            component: InputParams
          }
        },
        outputParams: {
          title: _r('输出参数'),
          type: 'object',
          properties: {},
          field: {
            component: OutputParams
          }
        },
        inputDefault: {
          title: '',
          type: 'object',
          properties: {}
        }
        // inputParams: {
        //   type: 'array',
        //   title: _r('输入参数'),
        //   items: {
        //     type: 'object',
        //     properties: {
        //       paramKey: {
        //         title: _r('参数名'),
        //         type: 'string',
        //         field: {
        //           validate: value => {
        //             const firstStr = value && value.slice(0, 1)
        //             if (/[0-9]/.test(firstStr)) {
        //               return _r('首位不可为数字')
        //             } else {
        //               return null
        //             }
        //           }
        //         }
        //       },
        //       variableType: {
        //         title: _r('参数类型'),
        //         type: 'object',
        //         field: {
        //           component: props => <C is='ParamTypeSchema' {...props} />
        //         }
        //       }
        //     }
        //   }
        // },
        // outputParams: {
        //   type: 'array',
        //   title: _r('输出参数'),
        //   items: {
        //     type: 'object',
        //     properties: {
        //       paramKey: {
        //         title: _r('参数名'),
        //         type: 'string',
        //         field: {
        //           validate: value => {
        //             const firstStr = value && value.slice(0, 1)
        //             if (/[0-9]/.test(firstStr)) {
        //               return _r('首位不可为数字')
        //             } else {
        //               return null
        //             }
        //           }
        //         }
        //       },
        //       variableType: {
        //         title: _r('参数类型'),
        //         type: 'object',
        //         field: {
        //           component: props => <C is='ParamTypeSchema' {...props} />
        //         }
        //       }
        //     }
        //   }
        // }
      },
      required: ['name', 'key', 'describe', 'runMode', 'apiKey', 'algorithmID', 'function'],
      permission: { add: true, edit: true, view: true, delete: true },
      components: {
        ListPage: MyListPage,
        EditPage: AddEditPage,
        AddPage: AddEditPage,
      }
    }
  },
  routers: {
    '/app/': [
      { path: 'algorithm/debug/:id', component: AlgorDebugPage },
    ]
  },
  icons: [
    { name: 'MyAlgorithm', url: require('../icons/我的算法.svg') },
  ],
  menuItems: [
    // { title: _r('AI算法库'), key: 'algorithm', icon: 'DataApi', url: '/app/model/algorithm/list', permission: 'algorithm.view' },
    { title: _r('我的算法'), key: 'algorithm', icon: 'MyAlgorithm', url: '/app/model/algorithm/list', permission: 'algorithm.view' }
  ]
}
