// @airiot/settings/src/settings/dashboardconfig.js
import PermissionConfig from './PermissionConfig'
import locales from '../../locales'

export default {
    name: 'iot-Dashboardconfig',
    locales,
    'settings.schema': app => ({ data: settings, ...props }, cb) => {
        const user = JSON.parse(sessionStorage.getItem('user'))
        if (user && user.isSuper) {
            cb(null, {
                name: 'dashboardconfig',
                title: _r('记录日志组件'),
                parentIndex: 'modular',
                type: 'object',
                properties: {
                    permission: {
                        type: 'array',
                        title: _r('权限'),
                        items: {}
                    }
                },
                form: ['*', { key: 'permission', component: PermissionConfig }]
            })
        } else {
            if (user && user.permissions && user.permissions.indexOf('dashboardConfig.view') > -1) {
                cb(null, {
                    name: 'dashboardconfig',
                    title: _r('记录日志组件'),
                    parentIndex: 'modular',
                    type: 'object',
                    properties: {
                        permission: {
                            type: 'array',
                            title: _r('权限'),
                            items: {}
                        }
                    },
                    form: ['*', { key: 'permission', component: PermissionConfig }]
                })
            } else {
                cb(null, {})
            }
        }
    }
}
