// @airiot/settings/src/settings/onlineSetting.js
import React from 'react'
import { api } from 'xadmin'
import { Checkbox, Button, message, Row, Col } from 'antd'
import _ from 'lodash'

const OnlineCheck = ({ data }) => {
  const [checked, setChecked] = React.useState(data.onlineCheck)
  const { t: _t1 } = useI18n()

  const handleChange = e => {
    let checked = e.target.checked
    setChecked(checked)
  }

  const handleClick = () => {
    let query = { onlineCheck: checked }
    api({ name: 'core/setting' })
      .fetch('', {
        method: 'PATCH',
        body: JSON.stringify(query)
      })
      .then(({ json }) => {
        message.success('系统设置保存成功！')
      })
      .catch(err => { message.error(err.message) })
  }

  return (
    <>
      <Row style={{ marginTop: '.6rem' }}>
        <Col span={6}></Col>
        <Col span={18}>
          <Checkbox onChange={handleChange} checked={checked}>{_t1('启用设备在线状态查询定时服务')}</Checkbox>
        </Col>
      </Row>
      <Row style={{ marginTop: '.6rem' }}>
        <Col span={6}></Col>
        <Col span={18}>
          <Button type="primary" onClick={() => handleClick()} style={{ margin: '0 .6rem' }}>{_t1('保存')}</Button>
          {/* <Button>{_t1('取消')}</Button> */}
        </Col>
      </Row>
    </>
  )
}

export default {
  name: 'iot-node-online-setting',
  'settings.schema': app => (props, cb) => {
    cb(null, {
      name: 'onlineCheck',
      title: _r('设备在线状态配置'),
      parentTab: _r('设备配置'),
      parentIndex: 'modular',
      renderComponent: props => <OnlineCheck {...props} />
    })
  }
}
