// @airiot/components/src/dashboard/AppPage.js
import React from 'react'
import { C } from 'xadmin-ui'

const AppContext = React.createContext()

const AppProvider = props => {
  return (
    <AppContext.Provider value={{ parentChildrens: props.parentChildrens }}>
      {props.children}
    </AppContext.Provider >
  )
}

const AppPage = {
  title: _r('内容区域'),
  category: [_r('通用组件'), _r('多级页面')],
  icon: require('../icons/前台页面.svg'),
  initLayout: { width: 550, height: 400 },
  component: props => {
    const { t:_t1 } = useI18n()
    const context = React.useContext(AppContext)
    if (props.dashboardMode && !context) {
      return (
        <div style={{ width: '100%', height: '100%', overflow: 'hidden' }}>
          <div style={{ display: 'flex', justifyContent: 'center', alignItems: 'center', width: '100%', height: '100%', minHeight: 130 }}>
            <C is="NoData" style={{ color: '#999', fontSize: 12, width: '100%', maxHeight: '100%' }} description={_t1("父级页面对应的画面中，需要添加该组件，用于子级页面的展示。")} />
          </div>
        </div>
      )
    }
    return context ? <>{context.parentChildrens}</> : null
  }
}

export default AppPage

export {
  AppProvider
}
