// @airiot/components/src/dashboard/ChangePassword.js
import React from 'react'
import { app, use, api } from 'xadmin'
import { Form } from 'xadmin-form'

import { Page, C } from 'xadmin-ui'
import { message } from 'antd';

const fields = [
  {
    name: 'old_password', attrs: { type: 'password' }, label: _r('原密码'), type: 'text',
    validate: value => (/^(?![^a-zA-Z]+$)(?!\D+$)/ig.test(value) ? null : _r('密码必须为字母与数字组合')),
  },
  {
    name: 'new_password', attrs: { type: 'password' }, label: '新密码', type: 'text',
    validate: value => (/^(?![^a-zA-Z]+$)(?!\D+$)/ig.test(value) ? null : _r('密码必须为字母与数字组合')),
  },
  {
    name: 'new_password2', attrs: { type: 'password' }, label: _r('再次输入新密码'), type: 'text',
    validate: (value, { new_password }) => value == new_password ? ((/^(?![^a-zA-Z]+$)(?!\D+$)/ig.test(value) ? null : _r('密码必须为字母与数字组合'))) : _r('两次输入密码不一致'),
  }
]

const ChangePassword = props => {

  const { t:_t1 } = useI18n()

  const onChange = async (values) => {
    try {
      await api({ name: 'core/user/changepwd' }).save(values)
      message.success(_t1('修改密码成功'))
      app.go('/login')
    } catch (err) {
      return (err.formError || err.json)
    }
  }

  return (
    <Page title={_t1("修改密码")}>
      <Form
        formKey="changePassword"
        fields={fields}
        onSubmit={onChange}
        component={C('Auth.ChangePassword') || C('Auth.Form')}
      />
    </Page>
  )
}

export default {
  title: _r('修改密码'),
  category: [_r('页面元素'), _r('导航栏')],
  icon: require('../icons/修改密码.svg'),
  component: ChangePassword,
  initLayout: { width: 380, height: 300 }
}
