// @airiot/components/src/dashboard/index.js
import LogTableWidget from './LogTableWidget'
import FrontMenu from './FrontMenu'
import UserName from './UserName'
import UserMenu from './UserMenu'
import ServiceTime from './ServiceTime'
import Profile from './Profile'
import ChangePassword from './ChangePassword'
import LogOut from './LogOut'
import AppPage, { AppProvider } from './AppPage'
import ScaleButton from './ScaleButton'
import LoadComponent from './Login/loadComponent'
import LogoutComponent from './Login/Logout'
import Reg from './Login/Reg'
import AuthInfoCard from './AuthInfoCard'
import PointTable from './PointTable'
import TableSelect from './TableSelect'
import TableDataSelect from './TableDataSelect'
import NavTabs from './NavTabs'
import TableRecordSelect from './TableRecordSelect'

export default {
  dashboardWidgets: {
    /* 前台组态组件 */
    AppPage,           // 子页面
    FrontMenu,         // 菜单
    UserMenu,          // 用户信息
    UserName,          // 用户名称
    Profile,           // 个人设置-限制只能修改基本信息
    ChangePassword,    // 修改密码 
    ServiceTime,       // 系统时间
    LogOut,            // 退出系统
    LogTableWidget,
    ScaleButton,
    LoadComponent,
    // Reg,
    LogoutComponent,
    PointTable,
    TableSelect,
    TableDataSelect,
    NavTabs,
    TableRecordSelect
    // AuthInfoCard
  },
  components: {
    'Dashboard.AppProvider': AppProvider
  },
  context: app => (context, cb) => {
    const FieldWrap = app.get('components')['Dashboard.Form.FieldWrap']
    if (FieldWrap) {
      app.get('dashboardWidgets')['PointTable'] = FieldWrap(PointTable)
      app.get('dashboardWidgets')['TableSelect'] = FieldWrap(TableSelect)
      app.get('dashboardWidgets')['TableDataSelect'] = FieldWrap(TableDataSelect)
    }
    cb(null, context)
  }
}
