// @airiot/components/src/dashboard/LogOut.js
import React from 'react'
import { use } from 'xadmin'
import { Button } from 'antd'

const LogOutComponent = (props) => {
  const { border, dashboardMode, text } = props
  const { onLogout } = use('auth.logout')
  const { t:_t1 } = useI18n()

  const style = { height: '100%' }

  if (border) {
    style['border'] = 'none'
    style['boxShadow'] = 'none'
  }

  return (
    <div className="dashboard-button-container" style={{ width: '100%', height: '100%' }}>
      <Button className="dashboard-button" onClick={!dashboardMode && onLogout} block style={style}>{text || _t1('退出系统')}</Button>
    </div>
  )
}

const paramSchema = {
  type: 'object',
  properties: {
    text: {
      title: _r('文字'),
      type: 'string'
    },
    border: {
      title: _r('隐藏边框'),
      type: 'boolean'
    },
  }
}

const LogOut = {
  title: _r('退出登录'),
  category: [_r('页面元素'), _r('登录')],
  icon: require('../icons/退出登录.svg'),
  component: LogOutComponent,
  initLayout: { width: 100, height: 80 },
  paramSchema,
  theme: [
    { selector: '.dashboard-button-container .dashboard-button', title: _r('退出登录样式及文字') },
    { selector: '.dashboard-button-container', title: _r('退出登录外层') }, 
  ]
}
export default LogOut
