// @airiot/components/src/dashboard/Profile.js
import { Card, message, Input } from 'antd';
import React from 'react';
import _ from 'lodash';
import { app, api } from 'xadmin';
import { Page, C, Loading } from 'xadmin-ui';
import { Model } from 'xadmin-model';

const UserBaseInfo = {
  'type': 'object',
  'name': 'user',
  'resource': 'user',
  'title': _r('个人资料'),
  'idProp': '_id',
  'key': 'user',
  'properties': {
    'username': {
      'title': _r('用户名'),
      'type': 'string',
      'field': {
        component: (props) => {
          const { input } = props
          return <Input disabled={true} value={input.value} />
        }
      }
    },
    'email': {
      'title': _r('邮件'),
      'type': 'string',
      'format': 'email'
    }
  },
  'partialSave': true,
  'permission': { add: true, edit: true, view: true, delete: true },
  'form': [
    'username',
    'email'
  ]
}

const ProfileComponent = props => {
  const [data, setData] = React.useState(null)

  const { t:_t1 } = useI18n()

  React.useEffect(() => {
    api({ name: 'core/auth/user' }).fetch('').then(({ json }) => {
      setData(json)
    })
  }, [])

  const saveSettings = (values) => {
    return api({ name: 'core/auth/user' }).fetch('', {
      method: 'PUT',
      body: JSON.stringify(_.omit(values, 'id'))
    })
      .then(() => message.success(_t1('保存个人设置成功')))
  }
 
  return (
    <Page title={_t1("个人设置")}>
      {data ? <Model schema={UserBaseInfo} modelKey="profile">
        <C is="Model.TabsForm" item={data} id={data.id} forceTabs mode="left" saveItem={saveSettings} />
      </Model> : <Loading />}
    </Page>
  )

}


const Profile = {
  title: _r('个人设置'),
  category: [_r('页面元素'), _r('导航栏')],
  icon: require('../icons/个人设置.svg'),
  component: ProfileComponent,
  initLayout: { width: 380, height: 300 }
}

export default Profile
