// @airiot/components/src/dashboard/TableRecordSelect.js
import React from 'react'
import { use, api } from 'xadmin'
import SelectTableRecord from '../components/SelectTableRecord'

const TableDataSelectComponent = (props) => {

  const { cellKey, field, placeholder, defaultValue, multiple, tables } = props
  const [ value, setValue ] = React.useState(defaultValue)
  const setOutputParams = use('cell.output.set', cellKey)
  const table = tables?.length ? tables.map(item => item.id) : null

  React.useEffect(() => {
    setOutputParams(defaultValue || {})
  }, [defaultValue])

  const onChange = (item) => {
    setValue(item)
    setOutputParams(item)
  }

  return (
    <div>
      {multiple ? 
      <SelectTableRecord  {...props} multiple={true} input={{ value: value instanceof Array ? value : value ? [value] : [], onChange}} field={{...field, placeholder}} table={table} /> :
      <SelectTableRecord {...props} input={{ value, onChange}} field={{...field, placeholder}} table={table} />}
    </div>
  )
}

const paramSchema = {
  type: 'object',
  properties: {
    tables: {
      type: 'array',
      title: '表选择',
      fieldType: 'table',
      items: {}
    },
    multiple: {
      type: 'boolean',
      title: _r('多选'),
    },
    placeholder: {
      title: _r('提示文字'),
      type: 'string',
      description: _r('选择器中的提示文本信息')
    },
    tree: {
      title: '树形展示',
      type: 'boolean'
    }
  }
}

const TableRecordSelect = {
  title: _r('记录选择器'),
  category: [_r('业务选择器'), _r('业务选择器')],
  icon: require('../icons/表记录选择器.svg'),
  component: TableDataSelectComponent,
  paramSchema: paramSchema,
  fieldWidget: true,
  initParam: { placeholder: _r('请选择记录') },
  initLayout: { width: 160, height: 40 }
}

export default TableRecordSelect
