// @airiot/components/src/dashboard/TableSelect.js
import React from 'react'
import { use, api } from 'xadmin'
import SelectTable from '../components/SelectTable'

const TableSelectComponent = (props) => {

  const { cellKey, field, placeholder, defaultValue, multiple } = props
  const [ value, setValue ] = React.useState(defaultValue)
  const setOutputParams = use('cell.output.set', cellKey)

  React.useEffect(() => {
    setOutputParams(defaultValue || {})
  }, [defaultValue])

  const onChange = (item) => {
    setValue(item)
    setOutputParams(item)
  }

  return (
    <div>
      {multiple ? 
      <SelectTable  {...props} mode="multiple" input={{ value: value instanceof Array ? value : value ? [value] : [], onChange}} field={{...field, placeholder}} /> :
      <SelectTable {...props} input={{ value, onChange}} field={{...field, placeholder}} />}
    </div>
  )
}

const paramSchema = {
  type: 'object',
  properties: {
    multiple: {
      type: 'boolean',
      title: _r('多选'),
    },
    placeholder: {
      title: _r('提示文字'),
      type: 'string',
      description: _r('选择器中的提示文本信息')
    },
    tree: {
      title: _r('树形展示'),
      type: 'boolean',
    }
  }
}

const TableSelect = {
  title: _r('表选择器'),
  category: [_r('业务选择器'), _r('业务选择器')],
  icon: require('../icons/表选择器.svg'),
  component: TableSelectComponent,
  paramSchema: paramSchema,
  fieldWidget: true,
  initParam: { placeholder: _r('请选择表') },
  initLayout: { width: 160, height: 40 }
}

export default TableSelect
