// @airiot/components/src/dashboard/Login/schema.js
import React from 'react'
import { ImageCode, Remember, Code } from './Code'
import { UserOutlined } from '@ant-design/icons';
import { Input } from 'antd'

const UserInput = (props) => {
  return (
    <>
      <div>username</div>
      <Input />
    </>
  )
}
//登录页面
const loginForm = {
  type: 'object',
  properties: {
    username: {
      type: 'string',
      title: _r('用户名')
    },
    password: {
      title: _r('密码'),
      type: 'string',
      maxLength: 30,
    },
    remenber: {
      // title: _r('记住我'),
      // type: 'boolean'
    },
    verifyCode: {
      title: _r('验证码'),
      type: 'string',
    }
  },
  required: ['username', 'password'],
  form: ['username', {
    key: 'password', attrs: { type: 'password' },
    // validate: value => (/^(?![^a-zA-Z]+$)(?!\D+$)/ig.test(value) ? null : '密码必须为字母与数字组合')
  },
    { key: 'verifyCode', component: props => <ImageCode {...props} /> },
    { key: 'remenber', component: Remember },
    //  { key: 'code', component: ImageCode } 
  ]
}
//注册页
const regForm = {
  type: 'object',
  properties: {
    com: {
      type: 'string',
      title: _r('公司')
    },
    username: {
      type: 'string',
      title: _r('用户名')
    },
    realname: {
      type: 'string',
      title: _r('姓名')
    },
    mobileNumber: {
      type: 'string',
      title: _r('手机号')
    },
    password: {
      type: 'string',
      title: _r('密码'),
      maxLength: 30,
      minLength: 6
    },
    email: {
      title: _r('邮箱'),
      type: 'string'
    }
  },
  form: [
    'username',
    {
      key: 'password', attrs: { type: 'password' },
      validate: value => (/^(?![^a-zA-Z]+$)(?!\D+$)/ig.test(value) ? null : _r('密码必须为字母与数字组合'))
    }, 'realname',
    { key: 'mobileNumber', validate: value => ((/^1[3456789]\d{9}$/).test(value) ? null : _r('手机号必须为11位数字')) }, 
    'com', 'email'],
  required: ['username', 'password', 'realname', 'mobileNumber', 'com', 'email']
}
const pwdPage = {
  type: 'object',
  properties: {
    pwd: {
      type: 'string',
      title: _r('重置密码')
    }
  },
  form: ['pwd']
}
//手机号重置密码
const resetPhoneForm = {
  type: 'object',
  properties: {
    phone: {
      type: 'string',
      title: _r('手机号')
    },
    verifyCode: {
      type: 'string',
      title: _r('验证码')
    }
  },
  form: ['phone', { key: 'verifyCode', component: ImageCode }]
}
//邮箱重置密码
const resetEmailForm = {
  type: 'object',
  properties: {
    email: {
      type: 'string',
      title: _r('邮箱')
    },
    verifyCode: {
      type: 'string',
      title: _r('验证码')
    }
  },
  form: ['email', 'verifyCode']
}

export {
  loginForm,
  regForm,
  pwdPage,
  resetPhoneForm,
  resetEmailForm
}
