// @airiot/components/src/Demo/fieldTypes.js
import { C } from 'xadmin-ui'

const schema = {
  type: 'object',
  properties: {
    dateTime: {
      title: _r('时间选择'),
      type: 'string',
      format: 'date-time'
    },
    department: {
      type: 'object',
      title: _r('部门'),
      properties: {
        departmentObject: {
          title: _r('部门单选'),
          type: 'object',
          fieldType: 'department'
        },
        departmentArray: {
          title: _r('部门多选'),
          type: 'array',
          fieldType: 'department',
          items: {}
        },
      }
    },
    table: {
      type: 'object',
      title: _r('数据表'),
      properties: {
        tableObject: {
          title: _r('数据表单选'),
          type: 'object',
          fieldType: 'table'
        },
        tableArrayId: {
          title: _r('数据表多选'),
          type: 'array',
          fieldType: 'table',
          items: {}
        }
      }
    },
    tableData: {
      type: 'object',
      title: '数据表记录(指定表范围)',
      properties: {
        tableAll: {
          title: _r('全部表'),
          type: 'object',
          fieldType: 'tableData'
        },
        tableOne: {
          title: _r('数据表单选'),
          type: 'object',
          fieldType: 'tableData',
          field: {
            table: '视频测试'
          }
        },
        tableMuti: {
          title: _r('数据表多选'),
          type: 'array',
          fieldType: 'tableData',
          field: {
            table: ['视频测试', '测试驱动jh']
          },
          items: {}
        },
        tableDataMuti: {
          title: _r('单表记录多选'),
          type: 'array',
          fieldType: 'tableData',
          items: {},
          field: {
            table: '视频测试'
          }
        },
      }
    },
    model: {
      type: 'object',
      title: _r('设备表'),
      description: '设备表只是func="device"的表',
      properties: {
        modelObject: {
          title: _r('设备表单选'),
          type: 'object',
          fieldType: 'table',
          field: {
            func: 'device',
          }
        },
        modelArray: {
          title: _r('设备表多选'),
          type: 'array',
          fieldType: 'table',
          field: {
            func: 'device'
          },
          items: {}
        }
      }
    },
    dept: {
      type: 'object',
      title: _r('部门表记录'),
      description: '设备表记录只是func="dataAuth"的表记录',
      properties: {
        nodeObject: {
          title: _r('部门表记录单选'),
          type: 'object',
          fieldType: 'tableData',
          field: {
            func: 'dataAuth'
          }
        },
        nodeArray: {
          title: _r('部门表记录多选'),
          type: 'array',
          fieldType: 'tableData',
          field: {
            func: 'dataAuth'
          },
          items: {}
        }
      }
    },
    node: {
      type: 'object',
      title: _r('设备表记录'),
      description: '设备表记录只是func="device"的表记录',
      properties: {
        nodeObject: {
          title: _r('设备表记录单选'),
          type: 'object',
          fieldType: 'tableData',
          field: {
            func: 'device'
          }
        },
        nodeArray: {
          title: _r('设备表记录多选'),
          type: 'array',
          fieldType: 'tableData',
          field: {
            func: 'device'
          },
          items: {}
        }
      }
    },
    dependent: {
      type: 'object',
      title: _r('表与记录联动'),
      properties: {
        tableObject: {
          title: _r('表单选'),
          type: 'object',
          properties: {
            table: {
              title: _r('表选择'),
              type: 'object',
              fieldType: 'table'
            },
            tableData: {
              title: _r('表记录选择'),
              type: 'array',
              fieldType: 'tableData',
              field: {
                depend: 'table' // schema key名
              },
              items: {},
              description: 'depend字段需要关联同级字段'
            },
          },
        },
        tableArray: {
          title: _r('表多选'),
          type: 'object',
          properties: {
            table: {
              title: _r('表选择'),
              type: 'array',
              items: {},
              fieldType: 'table'
            },
            tableData: {
              title: _r('表记录选择'),
              type: 'object',
              fieldType: 'tableData',
              field: {
                depend: 'table' // schema key名
              },
              description: 'depend字段需要关联同级字段'
            }
          }
        },
        arr: {
          title: _r('数组示例'),
          type: 'array',
          items: {
            type: 'object',
            properties: {
              table: {
                title: _r('表选择'),
                type: 'string',
                fieldType: 'table'
              },
              tableData: {
                title: '数据表单选（返回id）',
                type: 'object',
                fieldType: 'tableData',
                field: {
                  depend: 'table' // schema key名
                },
                description: 'depend字段需要关联同级字段'
              }
            }
          }
        }
      }
    },
    tagsSelect: {
      type: 'string',
      title: _r('指定表和表记录选择数据点'),
      fieldType: 'tags-select',
      field: {
        tableId: 'wxhtest',
        tableDataId: '6340daf099ea21544bcc1300'
      }
    }
  }
}

const Demo = () => {

  const onChange = (values) => {
    console.log(values, 'values')
  }

  return <C is="I18nSchemaForm"
    schema={schema}
    initialValues={{}}
    onChange={onChange}
  />
}

export default Demo
