// @airiot/warning/src/apps/warning/model_record.js
import { Icon, C, Loading } from 'xadmin-ui';
import { HasPermission } from 'xadmin-auth';
import { Button } from 'antd'
import { app }  from 'xadmin'

export default {
  tRecord: {
    title: _r('设备配置信息'),
    type: 'object',
    initQuery: false,
    userPageFields: true,
    noLocalStorageFilter: true,
    icon: <Icon svg={require('./../../icon/warning.svg')} />,
    name: 'record',
    resource: 'core/t/record',
    properties: {
      tableInfo: {
        title: _r('设备表'),
        type: 'object',
        fieldType: 'table',
        properties: {},
      },
      tableData: {
        title: _r('设备'),
        type: 'object',
        fieldType: 'tableData',
        properties: {}
      },
      status: {
        title: _r('状态'),
        type: 'string',
        width: 80,
      },
      connectTime: {
        title: _r('通信时间'),
        type: 'string',
        width: 250,
        format: 'date-time',
      },
      online: {
        title: _r('在线'),
        type: 'boolean',
        width: 80,
        boolLabel: ['在线', '掉线']
      }
    },
    listFields: [ 'tableInfo', 'tableData', 'status', 'connectTime', 'online' ],
    orders: { time: 'DESC' },
    fieldRender: {
      tableData: ({ item }) => {
        const { t:_t1, i18n } = useI18n()
        return item?.tableData?.name || item?.tableData?.id || _t1('空')
      },
      tableInfo: ({ item }) => {
        const { t:_t1, i18n } = useI18n()
        return item?.tableInfo?.title || item?.tableInfo?.name || item?.tableInfo?.id || _t1('空')
      },
    },
    form: ['*'],
    itemActions: [
      // ({ item }) => <Button onClick={() => app.go(`/app/model/node/${item.id}/view`)}>画面</Button>,
     'delete'
    ]
  }
}
