// @airiot/echarts/src/charts/dataset.js
const labelRight = {
  position: 'right'
};



export const defaultChartData = [
  {
    dimensions: ['dept', 'age'],
    source: [
      { dept: '10', age: '4.0' },
      { dept: '8.9', age: '5.7' },
      { dept: '6.7', age: '7.7' },
      { dept: '8.1', age: '8.6' },
      { dept: '4.9', age: '9.8' },
      { dept: '6.7', age: '6.8' },
      { dept: '7.8', age: '7.5' },
      { dept: '4.6', age: '9.7' },
      { dept: '8.4', age: '7.9' },
    ]
  },
  {
    dimensions: ['dept', 'age'],
    source: [
      { dept: '10', age: '4.0' },
      { dept: '8.9', age: '5.7' },
      { dept: '6.7', age: '7.7' },
      { dept: '8.1', age: '8.6' },
      { dept: '4.9', age: '9.8' },
      { dept: '6.7', age: '6.8' },
      { dept: '7.8', age: '7.5' },
      { dept: '4.6', age: '9.7' },
      { dept: '8.4', age: '7.9' },
    ]
  },
  {
    dimensions: ['dept', 'age'],
    source: [
      { dept: '10', age: '4.0' },
      { dept: '8.9', age: '5.7' },
      { dept: '6.7', age: '7.7' },
      { dept: '8.1', age: '8.6' },
      { dept: '4.9', age: '9.8' },
      { dept: '6.7', age: '6.8' },
      { dept: '7.8', age: '7.5' },
      { dept: '4.6', age: '9.7' },
      { dept: '8.4', age: '7.9' },
    ]
  },
  {
    dimensions: ['dept', 'age'],
    source: [
      { dept: '10', age: '4.0' },
      { dept: '8.9', age: '5.7' },
      { dept: '6.7', age: '7.7' },
      { dept: '8.1', age: '8.6' },
      { dept: '4.9', age: '9.8' },
      { dept: '6.7', age: '6.8' },
      { dept: '7.8', age: '7.5' },
      { dept: '4.6', age: '9.7' },
      { dept: '8.4', age: '7.9' },
    ]
  },
  {
    indicator: [{
      text: '科技创新联盟',
      max: 100
    },
    {
      text: '科技成果转化中心',
      max: 100
    },
    {
      text: '工程技术中心（工程实验室）',
      max: 100
    },
    {
      text: '新型研发机构',
      max: 100
    },
    {
      text: '科技创孵化器',
      max: 100
    }
    ],
    radardata: [
      [85, 65, 55, 90, 82],
      [50, 20, 45, 30, 75],
      [37, 80, 12, 50, 25]
    ]
  },
  {
    indicator: [{
      "name": "N[0]",
      "max": 88
    }, {
      "name": "",
      "max": 88
    }, {
      "name": "W[270]",
      "max": 88
    }, {
      "name": "",
      "max": 88
    }, {
      "name": "S[180]",
      "max": 88
    }, {
      "name": "",
      "max": 88
    }, {
      "name": "E[90]",
      "max": 88
    }, {
      "name": "",
      "max": 88
    }],
    radardata: [60, 10, 20, 50, 60, 90, 60, 60]
  },
  {
    source: [
      ['quarter', '一季度', '二季度', '三季度', '四季度'],
      ['数据1', 4, 25, 8, 10],
      ['数据2', 10, 16, 12, 10]
    ]
  },
  //柱状图3 数据7
  {
    dimensions: ["day", 'Mon', 'Tue', 'Wed'],
    source: [
      ["1月", 320, 302, 301],
      ["2月", 120, 132, 101],
      ["3月", 220, 182, 191],
      ["4月", 150, 212, 201],
    ]
  },
  //仪表盘1,10数据8
  {
    name: 'SCORE',
    value: 70,
  },
  //仪表盘4数据9
  {
    value: 0.7,
    name: 'Grade Rating'
  },
  {
    value: 20,
    name: 'Good',
    title: {
      offsetCenter: ['-40%', '80%']
    },
    detail: {
      offsetCenter: ['-40%', '95%']
    }
  },
  {
    value: 40,
    name: 'Better',
    title: {
      offsetCenter: ['0%', '80%']
    },
    detail: {
      offsetCenter: ['0%', '95%']
    }
  },
  {
    value: 60,
    name: 'Perfect',
    title: {
      offsetCenter: ['40%', '80%']
    },
    detail: {
      offsetCenter: ['40%', '95%']
    }
  },
  //line条形图 大部分 除5,6,7,8数据13
  {
    source: [
      [null, 'A', 'B'],
      ['一月', 620, 280],
      ['二月', 732, 932],
      ['三月', 631, 291],
      ['四月', 534, 234],
      ['五月', 830, 290],
      ['六月', 630, 330]
    ]
  },
  //line条形图第5个数据14
  {
    dimensions: ["day", 'Mon', 'Tue', 'Wed'],
    source: [
      ["1月", 320, 302, 301],
      ["2月", 120, 132, 101],
      ["3月", 220, 182, 191],
      ["4月", 150, 212, 201],
    ]
  },
  //line条形图第8个数据15
  {
    dimensions: ['day', 'Precipitation', 'Temperature'],
    source: [
      ['Mon', 100, 5],
      ['Tue', 200, 3],
      ['Wed', 150, 15],
      ['Thu', 160, 9],
      ['Fri', 90, 2],
      ['Stu', 80, 19],
    ]
  },
  //饼图数据 16
  {
    source: [
      ['name', 'value'],
      ['面包', 13],
      ['巧克力', 33],
      ['饼干', 19],
      ['糖', 23],
    ]
  },
  //饼图数据 17
  {
    source: [
      { name: '面包', value: 13, per: 15 },
      { name: '巧克力', value: 33, per: 37 },
      { name: '饼干', value: 19, per: 22 },
      { name: '糖', value: 23, per: 26 },
    ]
  },
  //饼图6,7,8数据18
  {
    value: 65
  },
  //条形图数据 19
  {
    name: ['南昌转运中心', '广州转运中心', '沈阳转运中心'],
    value: [20, 50, 90],
    valdata: [683, 234, 578],
    myColor: ['#1089E7', '#F57474', '#34da62']
  },
  //条形图数据 20
  {
    source: [
      [null, 'A', 'B'],
      ['一月', 620, 280],
      ['二月', 732, 932],
      ['三月', 631, 291],
      ['四月', 534, 234],
      ['五月', 830, 290],
      ['六月', 630, 330]
    ]
  },
  //条形图数据 21
  {
    source: [
      { name: '面包', value: -13, label: labelRight },
      { name: '巧克力', value: 33 },
      { name: '饼干', value: -19, label: labelRight },
      { name: '糖', value: 23 },
      { name: '奶茶', value: 23 },
      { name: '炸鸡', value: - 23, label: labelRight },
    ]
  },
  //条形图10数据 22
  {
    xData: ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'],
    data: [220, 182, 191, 234, 290, 330, 310]
  },
  //条形图11 数据23
  {
    barTopColor: ['#02c3f1', '#53e568', '#02c3f1', '#53e568', '#02c3f1', '#53e568', '#02c3f1', '#53e568',],
    barBottomColor: [
      'rgba(2,195,241,0.1)',
      'rgba(83, 229, 104, 0.1)',

      'rgba(2,195,241,0.1)',
      'rgba(83, 229, 104, 0.1)',

      'rgba(2,195,241,0.1)',
      'rgba(83, 229, 104, 0.1)',

      'rgba(2,195,241,0.1)',
      'rgba(83, 229, 104, 0.1)',
    ],
    name: ['香蕉', '苹果', '草莓', '樱桃', '巧克力', '大枣', '冰激凌'],
    value: [22, 33, 55, 16, 43, 29, 40]
  },
  //仪表盘7 数据24
  {
    percent: 22,
    color_percent0: '',
    color_percent100: '',
    dotArray: [],
  },
  //散点图3 数据25
  {
    source: [
      [1, 4862.4],
      [2, 5294.7],
      [3, 5934.5],
      [4, 7171.0],
      [5, 8964.4],
      [6, 10202.2],
      [7, 11962.5],
      [8, 14928.3],
      [9, 16909.2],
      [10, 18547.9],
      [11, 21617.8],
      [12, 26638.1],
      [13, 34634.4],
      [14, 46759.4],
      [15, 58478.1],
      [16, 67884.6],
      [17, 74462.6],
      [18, 79395.7]
    ]
  },
  //水球图数据  26
  {
    value: [0.5, 0.5, 0.5]
  },
  //和地图配套的柱状图数据27
  {
    name: ['辽宁省', '黑龙江省', '吉林省', '内蒙古自治区', '河南省', '河北省', '西藏自治区'],
    value: [33, 44, 32, 19, 41, 25, 37]

  },
  //散点图4数据 28
  {
    value: [
      [161.2, 51.6], [167.5, 59.0], [159.5, 49.2], [157.0, 63.0], [155.8, 53.6],
      [170.0, 59.0], [159.1, 47.6], [166.0, 69.8], [176.2, 66.8], [160.2, 75.2],
      [172.5, 55.2], [170.9, 54.2], [172.9, 62.5], [153.4, 42.0], [160.0, 50.0],
      [147.2, 49.8], [168.2, 49.2], [175.0, 73.2], [157.0, 47.8], [167.6, 68.8],
      [159.5, 50.6], [175.0, 82.5], [166.8, 57.2], [176.5, 87.8], [170.2, 72.8],
      [174.0, 54.5], [173.0, 59.8], [179.9, 67.3], [170.5, 67.8], [160.0, 47.0],
      [154.4, 46.2], [162.0, 55.0], [176.5, 83.0], [160.0, 54.4], [152.0, 45.8],
      [162.1, 53.6], [170.0, 73.2], [160.2, 52.1], [161.3, 67.9], [166.4, 56.6],
      [168.9, 62.3], [163.8, 58.5], [167.6, 54.5], [160.0, 50.2], [161.3, 60.3],
      [167.6, 58.3], [165.1, 56.2], [160.0, 50.2], [170.0, 72.9], [157.5, 59.8],
      [167.6, 61.0], [160.7, 69.1], [163.2, 55.9], [152.4, 46.5], [157.5, 54.3],
      [168.3, 54.8], [180.3, 60.7], [165.5, 60.0], [165.0, 62.0], [164.5, 60.3],
      [156.0, 52.7], [160.0, 74.3], [163.0, 62.0], [165.7, 73.1], [161.0, 80.0],
      [162.0, 54.7], [166.0, 53.2], [174.0, 75.7], [172.7, 61.1], [167.6, 55.7],
      [151.1, 48.7], [164.5, 52.3], [163.5, 50.0], [152.0, 59.3], [169.0, 62.5],
      [164.0, 55.7], [161.2, 54.8], [155.0, 45.9], [170.0, 70.6], [176.2, 67.2],
      [170.0, 69.4], [162.5, 58.2], [170.3, 64.8], [164.1, 71.6], [169.5, 52.8],
      [163.2, 59.8], [154.5, 49.0], [159.8, 50.0], [173.2, 69.2], [170.0, 55.9],
      [161.4, 63.4], [169.0, 58.2], [166.2, 58.6], [159.4, 45.7], [162.5, 52.2],
      [159.0, 48.6], [162.8, 57.8], [159.0, 55.6], [179.8, 66.8], [162.9, 59.4],
      [161.0, 53.6], [151.1, 73.2], [168.2, 53.4], [168.9, 69.0], [173.2, 58.4],
      [171.8, 56.2], [178.0, 70.6], [164.3, 59.8], [163.0, 72.0], [168.5, 65.2],
      [166.8, 56.6], [172.7, 105.2], [163.5, 51.8], [169.4, 63.4], [167.8, 59.0],
      [159.5, 47.6], [167.6, 63.0], [161.2, 55.2], [160.0, 45.0], [163.2, 54.0],
      [162.2, 50.2], [161.3, 60.2], [149.5, 44.8], [157.5, 58.8], [163.2, 56.4],
      [172.7, 62.0], [155.0, 49.2], [156.5, 67.2], [164.0, 53.8], [160.9, 54.4],
      [162.8, 58.0], [167.0, 59.8], [160.0, 54.8], [160.0, 43.2], [168.9, 60.5],
      [158.2, 46.4], [156.0, 64.4], [160.0, 48.8], [167.1, 62.2], [158.0, 55.5],
      [167.6, 57.8], [156.0, 54.6], [162.1, 59.2], [173.4, 52.7], [159.8, 53.2],
      [170.5, 64.5], [159.2, 51.8], [157.5, 56.0], [161.3, 63.6], [162.6, 63.2],
      [160.0, 59.5], [168.9, 56.8], [165.1, 64.1], [162.6, 50.0], [165.1, 72.3],
      [166.4, 55.0], [160.0, 55.9], [152.4, 60.4], [170.2, 69.1], [162.6, 84.5],
      [170.2, 55.9], [158.8, 55.5], [172.7, 69.5], [167.6, 76.4], [162.6, 61.4],
      [167.6, 65.9], [156.2, 58.6], [175.2, 66.8], [172.1, 56.6], [162.6, 58.6],
      [160.0, 55.9], [165.1, 59.1], [182.9, 81.8], [166.4, 70.7], [165.1, 56.8],
      [177.8, 60.0], [165.1, 58.2], [175.3, 72.7], [154.9, 54.1], [158.8, 49.1],
      [172.7, 75.9], [168.9, 55.0], [161.3, 57.3], [167.6, 55.0], [165.1, 65.5],
      [175.3, 65.5], [157.5, 48.6], [163.8, 58.6], [167.6, 63.6], [165.1, 55.2],
      [165.1, 62.7], [168.9, 56.6], [162.6, 53.9], [164.5, 63.2], [176.5, 73.6],
      [168.9, 62.0], [175.3, 63.6], [159.4, 53.2], [160.0, 53.4], [170.2, 55.0],
      [162.6, 70.5], [167.6, 54.5], [162.6, 54.5], [160.7, 55.9], [160.0, 59.0],
      [157.5, 63.6], [162.6, 54.5], [152.4, 47.3], [170.2, 67.7], [165.1, 80.9],
      [172.7, 70.5], [165.1, 60.9], [170.2, 63.6], [170.2, 54.5], [170.2, 59.1],
      [161.3, 70.5], [167.6, 52.7], [167.6, 62.7], [165.1, 86.3], [162.6, 66.4],
      [152.4, 67.3], [168.9, 63.0], [170.2, 73.6], [175.2, 62.3], [175.2, 57.7],
      [160.0, 55.4], [165.1, 104.1], [174.0, 55.5], [170.2, 77.3], [160.0, 80.5],
      [167.6, 64.5], [167.6, 72.3], [167.6, 61.4], [154.9, 58.2], [162.6, 81.8],
      [175.3, 63.6], [171.4, 53.4], [157.5, 54.5], [165.1, 53.6], [160.0, 60.0],
      [174.0, 73.6], [162.6, 61.4], [174.0, 55.5], [162.6, 63.6], [161.3, 60.9],
      [156.2, 60.0], [149.9, 46.8], [169.5, 57.3], [160.0, 64.1], [175.3, 63.6],
      [169.5, 67.3], [160.0, 75.5], [172.7, 68.2], [162.6, 61.4], [157.5, 76.8],
      [176.5, 71.8], [164.4, 55.5], [160.7, 48.6], [174.0, 66.4], [163.8, 67.3]
    ]
  },
  //地图经纬度数据 29
  {
    geoCoordMap : {
      '台湾省': [121.5135, 25.0308],
      '黑龙江省': [127.9688, 45.368],
      '内蒙古自治区': [110.3467, 41.4899],
      "吉林省": [125.8154, 44.2584],
      '北京市': [116.4551, 40.2539],
      "辽宁省": [123.1238, 42.1216],
      "河北省": [114.4995, 38.1006],
      "天津市": [117.4219, 39.4189],
      "山西省": [112.3352, 37.9413],
      "陕西省": [109.1162, 34.2004],
      "甘肃省": [103.5901, 36.3043],
      "宁夏回族自治区": [106.3586, 38.1775],
      "青海省": [101.4038, 36.8207],
      "新疆维吾尔自治区": [87.9236, 43.5883],
      "西藏自治区": [91.11, 29.97],
      "四川省": [103.9526, 30.7617],
      "重庆市": [108.384366, 30.439702],
      "山东省": [117.1582, 36.8701],
      "河南省": [113.4668, 34.6234],
      "江苏省": [118.8062, 31.9208],
      "安徽省": [114.3896, 30.6628],
      "浙江省": [119.5313, 29.8773],
      "福建省": [119.4543, 25.9222],
      "江西省": [116.0046, 28.6633],
      "湖南省": [106.6992, 26.7682],
      "云南省": [102.9199, 25.4663],
      "广东省": [113.12244, 23.009505],
      "广西壮族自治区": [108.479, 23.1152],
      "海南省": [110.3893, 19.8516],
      '上海市': [121.4648, 31.2891],
      "湖北省": [114.3896, 30.6628],
      "贵州省": [106.6992, 26.7682],
    }
  },
  //地图1 气泡数据 30
  {
    mapData : [
      { name: "天津市", value: 42 },
      { name: "北京市", value: 199 },
      { name: "河北省", value: 102 },
      { name: "山西省", value: 81 },
      { name: "内蒙古自治区", value: 47 },
      { name: "辽宁省", value: 67 },
      { name: "吉林省", value: 82 },
      { name: "黑龙江省", value: 123 },
      { name: "上海市", value: 24 },
      { name: "江苏省", value: 92 },
      { name: "浙江省", value: 114 },
      { name: "安徽省", value: 109 },
      { name: "福建省", value: 116 },
      { name: "江西省", value: 91 },
      { name: "山东省", value: 119 },
      { name: "河南省", value: 137 },
      { name: "湖北省", value: 116 },
      { name: "湖南省", value: 114 },
      { name: "重庆市", value: 91 },
      { name: "四川省", value: 125 },
      { name: "贵州省", value: 62 },
      { name: "云南省", value: 83 },
      { name: "西藏自治区", value: 9 },
      { name: "陕西省", value: 80 },
      { name: "甘肃省", value: 56 },
      { name: "青海省", value: 10 },
      { name: "宁夏回族自治区", value: 18 },
      { name: "新疆维吾尔自治区", value: 180 },
      { name: "广东省", value: 123 },
      { name: "广西壮族自治区", value: 59 },
      { name: "海南省", value: 14 },
      { name: "台湾省", value: 104 },
    ]
  },
  //地图2 数据31
  {
    chinaDatas : [
      [{
        name: '黑龙江省',
        value: 0
      }], [{
        name: '内蒙古自治区',
        value: 0
      }], [{
        name: '吉林省',
        value: 0
      }], [{
        name: '辽宁省',
        value: 0
      }],
      [{
        name: '台湾省',
        value: 0
      }],
      [{
        name: '河北省',
        value: 0
      }],
      [{
        name: '天津市',
        value: 0
      }], [{
        name: '山西省',
        value: 0
      }], [{
        name: '陕西省',
        value: 0
      }], [{
        name: '甘肃省',
        value: 0
      }], [{
        name: '宁夏回族自治区',
        value: 0
      }], [{
        name: '青海省',
        value: 0
      }], [{
        name: '新疆维吾尔自治区',
        value: 0
      }], [{
        name: '西藏自治区',
        value: 0
      }], [{
        name: '四川省',
        value: 0
      }], [{
        name: '重庆市',
        value: 0
      }], [{
        name: '山东省',
        value: 0
      }], [{
        name: '河南省',
        value: 0
      }], [{
        name: '江苏省',
        value: 0
      }], [{
        name: '安徽省',
        value: 0
      }], [{
        name: '湖北省',
        value: 0
      }], [{
        name: '浙江省',
        value: 0
      }], [{
        name: '福建省',
        value: 0
      }], [{
        name: '江西省',
        value: 0
      }], [{
        name: '湖南省',
        value: 0
      }], [{
        name: '贵州省',
        value: 0
      }], [{
        name: '广西壮族自治区',
        value: 0
      }], [{
        name: '海南省',
        value: 0
      }], [{
        name: '上海市',
        value: 1
      }],
      [{
        name: '云南省',
        value: 1
      }]
    ]
  }
]
