// @airiot/warning/src/Store/index.js
import _ from 'lodash'
import React from 'react'
import { use } from 'xadmin'
import {
  RecoilRoot,
  atom,
  atomFamily,
  selector,
  useRecoilValue,
  useSetRecoilState,
  useRecoilSnapshot,
} from 'recoil'

const Store = ({ id, data, children, mode = false }) => {
  return (
    <RecoilRoot>
      {children}
    </RecoilRoot>
  )
}

const warningUnread = atom({ key: 'warningUnreadAtom', default: [] })

const warningWsNewUnread = atom({ key: 'warningWsNewUnreadAtom', default: [] })

const warningtypeCount = atom({ key: 'warningStatAtom', default: 0 })

const warningProcessedCount = atom({ key: 'warningProcessedCountAtom', default: 0 }) 

const warningCount = atom({ key: 'warningCountAtom', default: 0 }) 

const warningView = atom({ key: 'warningCountAtom', default: 0 }) 

const initWarning = atom({ key: 'initWarningAtom', default: [] })

export default Store

export {
  warningUnread,
  warningWsNewUnread,
  warningtypeCount,
  warningProcessedCount,
  warningCount,
  warningView,
  initWarning
}
