// @airiot/settings/src/component/RouterEditTree.js
import React from 'react';
import _ from 'lodash'
import reduce from 'async/reduce';
import { Tree, Button, Modal, Select, Card, Form, Row, Col, TreeSelect, message } from 'antd';
import { SchemaForm } from 'xadmin-form';
import { Icon, C, Loading } from 'xadmin-ui';
import { _t } from 'xadmin-i18n';
import { app, use, api } from 'xadmin';
import IconPicker from './IconPicker'
import './MenuEditTree.css';
import { HasPermission } from 'xadmin-auth'

const { TreeNode } = Tree;
const { Option } = Select;

const travelTree = ts => {
  return ts ? ts.map(t => ({
    selectable: Boolean(t.url),
    ...t,
    ...(t.children ? { children: travelTree(t.children) } : {}),
    value: t.value || t.key || t.url || t.title,
  })) : null
}

const findTree = (ts, value) => {
  for (let i = 0; i < ts.length; i++) {
    const t = ts[i]
    if (t.value == value) {
      return t
    } else if (t.children) {
      const t1 = findTree(t.children, value)
      if (t1) {
        return t1
      }
    }
  }
  return null
}

const DashboardMenuItemSelect = ({ input }) => {
  const { form } = use('form')

  const [data, setData] = React.useState([])

  React.useEffect(() => {
    api('Dashboard')
      .query({ fields: ['name'] }, { where: { isDelete: { "$ne": true } } })
      .then(({ items }) => {
        setData(items)
      })
  }, [])

  const onChange = (v) => {
    input.onChange(v)
  }
  return (
    <Select
      showSearch
      optionFilterProp="children"
      filterOption={(input, option) =>
        option.children.toLowerCase().indexOf(input.toLowerCase()) >= 0
      }
      style={{ width: '100%' }}  {...input} onChange={onChange}
    >
      {data && data.map((item) => (<Option key={item.id} value={'dashboard.' + item.id}>{item.name}</Option>))}
    </Select>
  )
}

const loop = (data, key, callback) => {
  data.forEach((item, index, arr) => {
    if (item.key == key) {
      return callback(item, index, arr);
    }
    if (item.children) {
      return loop(item.children, key, callback);
    }
  });
};

const treeMap = (data, key, callback) => {
  return data.map((item, index, arr) => {
    if (item.key == key) {
      return callback(item, index, arr);
    }
    if (item.children) {
      return {
        ...item,
        children: treeMap(item.children, key, callback)
      }
    }
    return item
  }).filter(Boolean);
};

const newKey = (v, item) => {
  return (Math.random() * 1000).toString()
}

let routerForm = {
  type: 'object',
  properties: {
    title: {
      type: 'string',
      title: _r('页面名称')
    },
    path: {
      type: 'string',
      title: _r('画面选择'),
    },
    path: {
      type: 'string',
      title: _r('页面路径'),
    },
    element: {
      type: 'string',
      title: _r('页面画面')
    },
    permission: {
      type: 'string',
      title: _r('页面权限')
    }
  },
  required: ['title', 'path', 'element'],
  form: ['title', { key: 'element', component: DashboardMenuItemSelect }, '*']
}

const FormLayout = ({ children, change, invalid, handleSubmit, submitting, onCancel, limits, handleCancel, flag }) => {
  const { formState: { values } } = use('form')
  const { t: _t1 } = useI18n()
  return (
    <Form onSubmit={handleSubmit}>
      {/* <C is="Form.FieldGroup" label={_t1('系统页面')} meta={{}} field={{}}>
        <MenuItemSelect change={change} values={values || ''} limits={limits} />
      </C> */}
      {children}
      <Form.Item wrapperCol={{
        xs: { span: 24, offset: 0 },
        sm: { span: 18, offset: 5 }
      }}>
        <HasPermission FailureComponent={null} permission={`frontRouterMenu.${flag}`}>
          <Button type="primary" onClick={handleSubmit} loading={submitting} disabled={invalid}>{_t1('确定')}</Button>{' '}
          <Button onClick={() => {
            onCancel()
            handleCancel && handleCancel()
          }}>{_t('Cancel')}</Button>
        </HasPermission>
      </Form.Item>
    </Form>
  )
}

const MenuEditTree = ({ input, limits, handleCancel, type }) => {
  const { t: _t1 } = useI18n()
  const { value, onChange } = input
  const [selectKey, setSelect] = React.useState(null)
  const [parentKey, setParentKey] = React.useState(null)
  const [title, setTitle] = React.useState(null)
  const update = app?.context?.recoil?.useDataSet('customVar')
  // const data = app?.context?.recoil?.useDataValue('customVar')
  //拖拽功能
  const onDrop = info => {
    const dropKey = info.node.props.eventKey; //拖拽元素到达的位置元素
    const dragKey = info.dragNode.props.eventKey; //被拖拽元素
    const dropPos = info.node.props.pos.split('-');
    const dropPosition =
      info.dropPosition - Number(dropPos[dropPos.length - 1]);

    const data = [...value];

    // Find dragObject
    let dragObj;
    loop(data, dragKey, (item, index, arr) => {
      arr.splice(index, 1);
      dragObj = item;
    });
    // console.log(dragObj)
    if (!info.dropToGap) {
      // Drop on the content
      loop(data, dropKey, item => {
        item.children = item.children || [];
        // where to insert 示例添加到尾部，可以是随意位置
        item.children.push(dragObj);
      });
    } else if (
      (info.node.props.children || []).length > 0 && // Has children
      info.node.props.expanded && // Is expanded
      dropPosition == 1 // On the bottom gap
    ) {
      loop(data, dropKey, item => {
        item.children = item.children || [];
        // where to insert 示例添加到尾部，可以是随意位置
        item.children.unshift(dragObj);
      });
    } else {
      let ar;
      let i;
      loop(data, dropKey, (item, index, arr) => {
        ar = arr;
        i = index;
      });
      if (dropPosition == -1) {
        ar.splice(i, 0, dragObj);
      } else {
        ar.splice(i + 1, 0, dragObj);
      }
    }
    onChange(data);
  };

  //点击树显示编辑区域(查看页面项)
  const onSelect = (selectedKeys, e) => {
    const { eventKey, title } = e.node && e.node.props;
    setSelect(eventKey)
    setTitle(title)
  };

  //删除页面项
  const deleteTreeNode = () => {
    const newData = treeMap(value, selectKey, () => null)
    setSelect(null)
    onChange(newData)
  };

  //添加子级页面
  const addMenu = (values) => {
    let newData = value
    if (parentKey == '') {
      // 添加主页面
      newData = [...value, { key: newKey(values), ...values }]
    } else {
      newData = treeMap(value, parentKey, (item) => {
        item.children = [
          ...(item.children || []),
          { key: newKey(values, item), ...values }
        ]
        return item
      })
    }

    onChange(newData)
    setParentKey(null)
    setSelect(null)
  };

  //修改后更新树结构
  const updateData = (values) => {
    const newData = treeMap(value, selectKey, (item) => ({ ...item, ...values }))
    newData && onChange(newData)
    // 关闭弹出框
    handleCancel && handleCancel()
  };

  const selectItem = React.useMemo(() => {
    if (selectKey) {
      let data = null
      loop(value, selectKey, (item) => {
        data = { ...item }
      })
      return data
    } else {
      return null
    }
  }, [value, selectKey])

  //树形结构子级页面
  const MenuDate = (data) =>
    data && data.map(item => {
      if (item.children && item.children.length) {
        return (
          <TreeNode
            key={item.key}
            title={item.title}
            url={item.url}
          >
            {MenuDate(item.children)}
          </TreeNode>
        );
      }
      return (
        <TreeNode
          key={item.key}
          title={item.title}
          url={item.url}
        />
      );
    });


  const CardTitle = (<div style={{ display: 'flex' }}>
    <span style={{ flex: 1 }}>{_t1('页面管理')}</span>
    <HasPermission FailureComponent={null} permission="frontRouterMenu.add">
      <Icon style={{ fontSize: 22 }} name="folder-add" onClick={() => setParentKey('')} />
    </HasPermission>
  </div>)

  routerForm.form = type == 'front' ? ['title', 'url', 'permission', {
    key: 'icon', component: IconPicker, category: 'icon'
  }] : routerForm.form
  return (
    <Row>
      <Col span={6}>
        <Card style={{ minHeight: 300 }} title={CardTitle} >
          {
            value && value.length > 0 ? <Tree className="draggable-tree"
              draggable showLine autoExpandParent defaultExpandAll
              onDrop={onDrop}
              onSelect={onSelect}
            >
              {MenuDate(value)}
            </Tree>
              : <C is="NoData" description={_t1('请添加页面')}
                style={{
                  position: 'absolute', top: '50%', right: '50%', marginTop: '-50px', marginRight: '-53px'
                }} />
          }
        </Card>
      </Col>
      <Col span={18}>
        <Card style={{ marginLeft: '1rem', height: '100%' }} title={selectKey != null ? (
          <>
            <HasPermission FailureComponent={null} permission="frontRouterMenu.add">
              <Button onClick={() => { setParentKey(selectKey) }} size="small"  >
                添加子页面
              </Button>
            </HasPermission>{' '}
            <HasPermission FailureComponent={null} permission="frontRouterMenu.delete">
              <Button onClick={() => Modal.confirm({
                title: '确定要删除该页面项吗？删除后不可撤回！', onOk: deleteTreeNode
              })} size="small">
                删除
              </Button>
            </HasPermission>
          </>
        ) : null} >
          {
            selectKey != null ? <C is="I18nSchemaForm"
              formKey={'routerForm.r' + selectKey}
              schema={routerForm}
              initialValues={selectItem}
              onSubmit={values => updateData(values)}
              onCancel={() => setSelect(null)}
              component={props => <FormLayout  {...props} limits={limits} handleCancel={handleCancel} flag={'edit'} />}
            /> : <C is="NoData" description={_t1('请选择或添加页面')} style={{
              position: 'absolute',
              top: '50%', right: '50%', marginTop: '-100px', marginRight: '-53px'
            }} />
          }

        </Card>
      </Col>
      <Modal
        title={parentKey == '' ? '添加页面' : '添加子页面'}
        destroyOnClose={true}
        footer={null}
        width={'40%'}
        visible={parentKey != null}
        onCancel={() => setParentKey(null)}
      >
        <C is="I18nSchemaForm"
          formKey={'routerForm.child.r' + parentKey}
          schema={routerForm}
          onSubmit={values => addMenu(values)}
          onCancel={() => setParentKey(null)}
          component={props => <FormLayout  {...props} limits={limits} flag={'add'} />}
        />
      </Modal>
    </Row>
  )
}

export default MenuEditTree;
