// @airiot/settings/src/component/SafeRequest.js
import React, { useEffect, useState } from 'react'
import { use, app, api, config } from 'xadmin'
import _ from 'lodash'
import { SchemaForm } from 'xadmin-form'
import { Popover, message } from 'antd'
import { C, Icon, Page, Loading } from 'xadmin-ui'

const schema = {
  type: 'object',
  properties: {
    safeRequest: {
      title: _r('是否发送安全请求'),
      type: 'boolean',
      description: _r('选中后发送请求的方式只有post、get')
    }
  },
  form: [ '*' ]
}

const SafeRequest = props => {
  const { dispatch } = use('redux')
  const { settings } = use('settings')

  const onSubmit = value => {
    api({ name: 'core/setting' })
      .fetch('', { method: 'PATCH', body: JSON.stringify(value) })
      .then(({ json }) => {
        message.success('系统设置保存成功')
        dispatch({ type: '@@iot/GET_SETTINGS' })
      })
  }

  return <Page title={<span>{_t1('安全请求')}</span>}>
    <C is="I18nSchemaForm"
      formKey={'safeRequest'+settings?.safeRequest}
      schema={schema}
      initialValues={settings}
      onSubmit={(val) => onSubmit(val)}
    />
  </Page>
}

export default SafeRequest
