// @airiot/echarts/src/charts/pictorialBar/index.js
import React from 'react'
import { C } from 'xadmin-ui'
import _ from 'lodash'
import useChart from '../../components/useChart'
import { initLayout, chartDafaultData, defaultSchema } from '../common/base'

const seriesOption = {
  base: {
    title: _r('基础'),
    type: 'object',
    properties: {
      name: {
        title: _r('名字'),
        type: 'string',
        description: _r('系列名称，用于tooltip的显示，legend 的图例筛选'),
      },
      color: {
        title: _r('柱条颜色'),
        type: 'string',
        fieldType: 'color'
      },
      barWidth: {
        title: _r('柱条宽度'),
        type: 'number'
      },
      barMinHeight: {
        title: _r('最小高度'),
        type: 'number'
      },
      barCategoryGap: {
        title: _r('柱间距离'),
        type: 'number'
      },
      barGap: {
        title: _r('系列柱距离'),
        type: 'number'
      },
      symbolClip: {
        title: _r('是否剪裁'),
        type: 'boolean'
      },
      symbolRepeat: {
        title: _r('元素重复'),
        type: 'boolean'
      },
      zlevel: {
        type: 'number',
        title: _r('图形zlevel值')
      }
    }
  },
}

const chart_pictorialBar = {
  title: _r('象形柱图'),
  category: [_r('数据图表'), _r('柱状图')],
  icon: require('../../svg/newIcons/象形柱图.svg'),
  component: props => useChart({ ...props }),
  initLayout,
  initParam: { option: {
    series: [{ type: 'pictorialBar' }],
    xAxis: { type: 'category', show: true, axisLine: { lineStyle: { color: '#333' } } },
    yAxis: { type: 'value', show: true, axisLine: { lineStyle: { color: '#333' } } },
    dataset: chartDafaultData
  } },
  echartOption: {
    seriesOption
  },
  paramSchema: {
    type: 'object',
    properties: {
      ...defaultSchema
    }
  }
}

export default chart_pictorialBar
