// @airiot/warning/src/apps/warning/pages/WarningLog.js
import React, { useMemo } from 'react'
import { Button, Modal } from 'antd'
import { Model } from 'xadmin-model'
import { C } from 'xadmin-ui'
import { use } from 'xadmin'

const WarningLog = ({ item }) => {
  const { t:_t1, i18n } = useI18n()
  const [ visible, setVisible ] = React.useState(false)

  const showModal = () => setVisible(true)

  const handleCancel = e => { setVisible(false) }

  const dom = useMemo(() => {
    if (item) {
      return (
        <>
          <Button onClick={showModal}>{_t1('记录')}</Button>
          <Modal
            title={_t1("报警操作记录")}
            visible={visible}
            width={'85%'}
            footer={[]}
            onCancel={handleCancel}
          >
            {visible ?
              <Model name="Log" modelKey={`warinng-log-${item.id}`} props={{
                batchActions: [],
                itemActions: [
                  ({ item }) => {
                    const navigate = use('navigate')
                    return (
                      <Button onClick={() => navigate(`/app/model/Log/${item?.id}/detail`)}>{_t1('查看详情')}</Button>
                    )
                  } ],
                initialValues: {
                  wheres: {
                    filters: {
                      type: { '$in': [ '报警确认', '报警处理', '操作记录' ] },
                      warningId: item && item.id
                    }
                  }
                }
              }}>
                <C is="Model.DataTable" />
              </Model> : null}
          </Modal>
        </>
      )
    }
  }, [ visible, item && item.id ])

  return dom ||  ''
}

export default WarningLog
