// @airiot/warning/src/dashboard/index.js
import React from 'react'
import Button from '../components/WarningButton'
import Audio from '../components/WarningAudio'
import Bar from '../components/WarningBar';
import Notification from './Notification'
import { HasPermission } from 'xadmin-auth'

const WarningButton = {
  title: _r('报警铃铛'),
  category: [_r('页面元素'), _r('报警元素')],
  icon: require('./../icon/报警铃铛.svg'),
  component: props => (<HasPermission FailureComponent={null} permission='warning.view' ><Button {...props} /></HasPermission>),
  initLayout: { width: 100, height: 100 },
  paramSchema: {
    type: 'object',
    properties: {
      warnTypeNum: {
        title: _r('报警数量'),
        type: 'string',
        enum: ['type', 'processed', 'all'],
        enum_title: [_r('报警分类数'), _r('需处理报警数'), _r('所有报警')]
      },
      isAlarm: {
        title: _r('报警中'),
        type: 'string',
        field: {
          uploadType: 'img',
          btnText: '媒体库'
        },
        fieldType: 'media_modal',
      },
      noAlarm: {
        title: _r('无报警'),
        type: 'string',
        field: {
          uploadType: 'img',
          btnText: '媒体库'
        },
        fieldType: 'media_modal'
      },
      width: {
        title: _r('宽度'),
        type: 'number',
        description: '默认宽度50'
      },
      height: {
        title: _r('高度'),
        description: '默认高度50',
        type: 'number',
      },
      offsetX: {
        title: _r('左右偏移'),
        type: 'number',
      },
      offsetY: {
        title: _r('上下偏移'),
        type: 'number',
      },
      badgeX: {
        title: _r('数字左右偏移'),
        type: 'number',
      },
      badgeY: {
        title: _r('数字上下偏移'),
        type: 'number',
      }
    }
  },
}

const WarningAudio = {
  title: _r('报警声音'),
  category: [ _r('页面元素'), _r('报警元素')],
  icon: require('./../icon/报警声音.svg'),
  component: props => (<HasPermission FailureComponent={null} permission='warning.view' ><Audio {...props} /></HasPermission>),
  initLayout: { width: 100, height: 100 },
}

const WarningBar = {
  title: _r('报警卡片'),
  category: [_r('页面元素'), _r('报警元素')],
  icon: require('./../icon/报警卡片.svg'),
  component: props => (<HasPermission FailureComponent={null} permission='warning.view'><Bar {...props} /></HasPermission>),
  initLayout: { width: 100, height: 100 },
}

const dashboardWidgets = {
  WarningButton, WarningAudio, WarningBar
}

export {
  dashboardWidgets
}
