// @airiot/editor/src/front.js
import app, { api } from 'xadmin'
import { lazy } from 'xadmin-ui'
import TestEditor from './tests/TestEditor5'
import apiList from './components/apiDocEditor/api/api'
import dashBoardApiList from './components/apiDocEditor/api/dashboardApi'
import './style.less'
import locales from '../locales'

const CodeEditor = lazy(() => import('./components/CodeEditor'))
const DataFieldEditor = lazy(() => import('./components/DataFieldEditor'))
const ApiDocEditor = lazy(() => import('./components/ApiDocEditor'))
const VariableEditor = lazy(() => import('./components/VariableEditor'))
const LogicEditor = lazy(() => import('./components/LogicEditor'))


export default app.use({
  name: 'iot.editor',
  locales,
  components: {
    CodeEditor,
    'CodeEditor.ModelLogicEditor': LogicEditor,
    'CodeEditor.DataFieldEditor': DataFieldEditor,
    'CodeEditor.ApiDocEditor': ApiDocEditor,
    'CodeEditor.VariableEditor': VariableEditor
  },
  functionList : {
    ...apiList,
    ...dashBoardApiList
  },
  routers: {
    '/app/' : {
      path: 'editor',
      breadcrumbName: '编辑器调试',
      component: TestEditor
    }
  }
})
