// @airiot/editor/src/index.js
import app, { api } from 'xadmin'
import { lazy } from 'xadmin-ui'
import TestEditor from './tests/TestEditor5'
import Chinese from './components/Chinese' // 汉化，不可删除
import apiList from './components/apiDocEditor/api/api'
import dashBoardApiList from './components/apiDocEditor/api/dashboardApi'
import hooks from './hooks'
import './style.less'
import locales from '../locales'

const CodeEditor = lazy(() => import('./components/CodeEditor'))
const JsonLogicEditor = lazy(() => import('./components/JsonLogicEditor'))
const FormulaLogic = lazy(() => import('./components/FormulaLogicEditor'))
const DataFieldEditor = lazy(() => import('./components/DataFieldEditor'))
const ApiDocEditor = lazy(() => import('./components/ApiDocEditor'))
const VariableEditor = lazy(() => import('./components/VariableEditor'))
const LogicEditor = lazy(() => import('./components/LogicEditor'))
const AddFunction = lazy(() => import('./components/apiDocEditor/AddFunction'))


export default app.use({
  name: 'iot.editor',
  locales,
  components: {
    CodeEditor,
    // JsonLogic编辑器
    'CodeEditor.JsonLogicEditor': JsonLogicEditor,
    // 老版公式编辑器
    'CodeEditor.FormulaLogicEditor': FormulaLogic,
    'CodeEditor.DataFieldEditor': DataFieldEditor,
    'CodeEditor.ApiDocEditor': ApiDocEditor,
    'CodeEditor.VariableEditor': VariableEditor,
    'CodeEditor.AddFunction': AddFunction,
  },
  schema_converter: [(f, schema, options) => {
    if (schema.fieldType == 'model-logic') {
      f.type = 'model_logic'
    }
    return f
  }],
  form_fields: {
    model_logic: {
      component: LogicEditor
    }
  },
  functionList : {
    ...apiList,
    ...dashBoardApiList
  },
  routers: {
    '/app/' : {
      path: 'editor',
      breadcrumbName: '编辑器调试',
      component: TestEditor
    }
  },
  hooks
})
